src/main/java/com/github/davidmoten/rx2/file/Files.java

```diff
@@ -101,35 +105,59 @@ private Files() {
      * @return Flowable of strings
      */
     private static Flowable<String> tailLines(File file, long startPosition, int chunkSize, Charset charset,
-            Flowable<?> events) {
+            Observable<?> events) {
         Preconditions.checkNotNull(file);
         Preconditions.checkNotNull(charset);
         Preconditions.checkNotNull(events);
-        return toLines(events.compose(x -> eventsToBytes(x, file, startPosition, chunkSize)), charset);
+        return toLines(eventsToBytes(events, file, startPosition, chunkSize), charset);
     }
 
-    private static Flowable<WatchEvent<?>> events(WatchService watchService, Scheduler scheduler, long intervalMs) {
+    private static Observable<WatchEvent<?>> events(WatchService watchService, Scheduler scheduler, long intervalMs) {
         Preconditions.checkNotNull(watchService, "watchService cannot be null");
         Preconditions.checkNotNull(scheduler, "scheduler cannot be null");
         Preconditions.checkArgument(intervalMs > 0, "intervalMs must be positive");
-        return Flowable.interval(intervalMs, TimeUnit.MILLISECONDS, scheduler) //
+        return Observable.interval(intervalMs, TimeUnit.MILLISECONDS, scheduler) //
                 .flatMap(x -> {
                     try {
                         WatchKey key = watchService.poll();
-                        if (key != null) {
-                            Flowable<WatchEvent<?>> r = Flowable.fromIterable(key.pollEvents());
+                        if (key != null && key.isValid()) {
+                            Observable<WatchEvent<?>> r = Observable.fromIterable(key.pollEvents());
                             key.reset();
                             return r;
                         } else {
-                            return Flowable.empty();
+                            return Observable.empty();
                         }
                     } catch (ClosedWatchServiceException e) {
                         // ignore
-                        return Flowable.empty();
+                        return Observable.empty();
                     }
                 });
     }
 
+    private static Observable<WatchEvent<?>> eventsBlocking(WatchService watchService) {
+        Preconditions.checkNotNull(watchService, "watchService cannot be null");
+        return Observable.<WatchEvent<?>, Queue<WatchEvent<?>>>generate(() -> new LinkedList<WatchEvent<?>>(), //
+                (q, emitter) -> {
+                    try {
+                        while (q.isEmpty()) {
+                            // blocking call
+                            WatchKey key = watchService.take();
+                            if (key.isValid()) {
+                                q.addAll(key.pollEvents());
+                            }
+                            key.reset();
+                        }
+                        emitter.onNext(q.poll());
```

Comment:
I'm not sure I understand this code. You're potentially adding multiple events to the list, but call `onNext` only on the first?

Reply:
Ah, nevermind, seems to be the correct usage of `generate`.